### DISCUSSION TITLE
Using FontAwesome in Nuxt 3

### DISCUSSION THREAD

**Discussion by kevzz1994** [nuxt/nuxt#16014] [Questions] [answered]
I have the following in `plugins/fontawesome.js`:

```js
import {defineNuxtPlugin} from "nuxt3/app";

import { library, config } from '@fortawesome/fontawesome-svg-core'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import {faBars, faCalendarDay, faLocationDot} from '@fortawesome/pro-solid-svg-icons'

// This is important, we are going to let Nuxt.js worry about the CSS
config.autoAddCss = false

// You can add your icons directly in this plugin. See other examples for how you
// can add other styles or just individual icons.
library.add(faCalendarDay, faLocationDot, faBars)

// Register the component globally
export default defineNuxtPlugin((nuxtApp) => {
    nuxtApp.vueApp.component('FontAwesomeIcon', FontAwesomeIcon)
})
```

And when running Nuxt I get the following:

"Named export 'config' not found. The requested module '@fortawesome/fontawesome-svg-core' is a CommonJS module, which may not support all module.exports as named exports." 

But during `npm run dev` it works just fine. Any suggestions? 

I also tried importing using the default export and same issue.

**Comment by ennioVisco** [comment]
I have pretty much the same exact code and it works for me.
Try selecting the right version of vue-fontawesome, I have in `package.json`:
```js
dependencies {
...
    "@fortawesome/vue-fontawesome": "^3.0.0-5",
...
}
```

**Comment by websaxophone** [comment, marked as answer]
I've resolved this issue by following [nuxt 3  documentation](https://v3.nuxtjs.org/concepts/esm/#what-kinds-of-problems-can-there-be). I've added the following configuration to `nuxt.config.ts`:
```js
export default defineNuxtConfig({

    build: {
        transpile: [
            '@fortawesome/fontawesome-svg-core',
            '@fortawesome/pro-solid-svg-icons',
            '@fortawesome/pro-regular-svg-icons',
            '@fortawesome/pro-light-svg-icons',
            '@fortawesome/free-brands-svg-icons'
        ]
    }
    ...
```

My dependencies are:
```js
  "dependencies": {
    "@fortawesome/fontawesome-svg-core": "^6.1.1",
    "@fortawesome/free-brands-svg-icons": "^6.1.1",
    "@fortawesome/pro-duotone-svg-icons": "^6.1.1",
    "@fortawesome/pro-light-svg-icons": "^6.1.1",
    "@fortawesome/pro-regular-svg-icons": "^6.1.1",
    "@fortawesome/pro-solid-svg-icons": "^6.1.1",
    "@fortawesome/vue-fontawesome": "^3.0.0-5",
    ...
```

Font awesome works fine, but the only problem is I can see the following messages in server console for all my icons:
`
Could not find one or more icon(s) { prefix: 'fas', iconName: 'phone' } {}
`
This issue is only reproducible after building and running app in production mode.
Still trying to resolve it... Any help is appreciated :)

**Reply by phojie** [reply]
just include '@fortawesome/vue-fontawesome' in your transpile build 😄 

```
 build: {
      transpile: [
        '@fortawesome/vue-fontawesome',
        '@fortawesome/fontawesome-svg-core',
        '@fortawesome/pro-solid-svg-icons',
        '@fortawesome/pro-regular-svg-icons',
        '@fortawesome/free-brands-svg-icons'
      ]
    }
```

**Reply by websaxophone** [reply]
Thank you, @Phojie ! Now Font Awesome works as expected! 😊

**Reply by jamaluddinrumi** [reply]
i'm experiencing this too 

and worked by following https://github.com/FortAwesome/vue-fontawesome/issues/215 by replacing 
`Could not find one or more icon(s) { prefix: 'fas', iconName: 'fa-solid fa-bars' } {}` 
with 
`<font-awesome-icon :icon="['fas', 'bars']" />`

```
package.json

"devDependencies": {
    "@fortawesome/fontawesome-svg-core": "^6.1.1",
    "@fortawesome/free-regular-svg-icons": "^6.1.1",
    "@fortawesome/free-solid-svg-icons": "^6.1.1",
    "@fortawesome/free-brands-svg-icons": "^6.1.1",
    "@fortawesome/vue-fontawesome": "^3.0.0-5",
```

------------------------------
- Operating System: `Darwin`
- Node Version:     `v16.14.2`
- Nuxt Version:     `3.0.0-rc.1`
- Package Manager:  `yarn@1.22.18`
- Builder:          `vite`
- User Config:      `typescript`, `publicRuntimeConfig`, `privateRuntimeConfig`, `css`, `buildModules`, `build`
- Runtime Modules:  `-`
- Build Modules:    `@nuxtjs/tailwindcss@5.0.3`, `@vueuse/nuxt@7.7.1`, `@nuxtjs/partytown@1.1.2`
------------------------------

**Reply by cawa-93** [reply]
You may not need any plugins, transpiles, configs. Look at [here](https://www.npmjs.com/package/@iconify-prerendered/vue-fa6-brands).
```vue
<script setup>
// Import components as usual
import {
    IconStrava,
    IconCss3,
    IconYCombinator
} from '@iconify-prerendered/vue-fa6-brands'
</script>

<template>
  <!-- And just use it in template -->
  <IconStrava/>
  <IconCss3/>
  <IconYCombinator/>
</template>
```

**Reply by Alqua** [reply]
> just include '@fortawesome/vue-fontawesome' in your transpile build 😄
> 
> ```
>  build: {
>       transpile: [
>         '@fortawesome/vue-fontawesome',
>         '@fortawesome/fontawesome-svg-core',
>         '@fortawesome/pro-solid-svg-icons',
>         '@fortawesome/pro-regular-svg-icons',
>         '@fortawesome/free-brands-svg-icons'
>       ]
>     }
> ```

Why is this not mention in the fontawesome doc? That fixed my issues!

**Reply by szq233** [reply]
> just include '@fortawesome/vue-fontawesome' in your transpile build 😄
> 
> ```
>  build: {
>       transpile: [
>         '@fortawesome/vue-fontawesome',
>         '@fortawesome/fontawesome-svg-core',
>         '@fortawesome/pro-solid-svg-icons',
>         '@fortawesome/pro-regular-svg-icons',
>         '@fortawesome/free-brands-svg-icons'
>       ]
>     }
> ```

wow，That fixed my issues too！Thank you！

**Reply by coolCucumber-cat** [reply]
Worked for me too, you just have to copy and paste the the paths of the imports in the plugin file 👍

**Reply by Trevor-143** [reply]
I have been working on this error for 3 days, on the 2nd day i broke my app fixing it with all sorts of methods now let me try this.

**Reply by Trevor-143** [reply]
weeeeeellllll, it worked so flawlessly. Thank you so much.

**Comment by hsyoo1219** [comment]
it's not work for me :(
plz let me know how to add fontawesome to nuxt3

`/clients/plugins/fontawesome.ts`
```
iimport { defineNuxtPlugin } from 'nuxt3/app'

import { library, config } from '@fortawesome/fontawesome-svg-core'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import { faArrowRotateRight } from '@fortawesome/free-solid-svg-icons'
library.add(faArrowRotateRight)

config.autoAddCss = false

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.vueApp.component('FontAwesomeIcon', FontAwesomeIcon)
})
```

`nuxt.config.ts`
```
import { defineNuxtConfig } from 'nuxt3'

export default defineNuxtConfig({
  plugins: [
    '~/clients/plugins/fontawesome.ts'
  ],
  build: {
    transpile: [
      '@fortawesome/vue-fontawesome',
      '@fortawesome/fontawesome-svg-core',
      '@fortawesome/free-regular-svg-icons',
      '@fortawesome/free-solid-svg-icons',
    ]
  }
})

```
`vue`

```
<template>
  <div>
    hello
    <font-awesome-icon icon="fa-solid fa-arrow-rotate-right" />
  </div>
</template>
```

I have error message like this
```
[Vue warn]: Failed to resolve component: font-awesome-icon
If this is a native custom element, make sure to exclude it from component resolution via compilerOptions.isCustomElement.
[nitro] [dev] [unhandledRejection] TypeError: Invalid value used as weak map key
[Vue warn]: Failed to resolve component: font-awesome-icon
```

**Reply by stafyniaksacha** [reply]
You can use [unocss](https://github.com/unocss/unocss) more easily:

```ts
// nuxt.config.ts
import presetIcons from '@unocss/preset-icons'
import { defineNuxtConfig } from 'nuxt3'

export default defineNuxtConfig({
  buildModules: ['@unocss/nuxt'],

  unocss: {
    preflight: false,
    presets: [
      presetIcons(),
    ],
  },
})
```

Then in any components:

```html
<template>
  <i class="i-fa-regular-heart" />
  <div class="i-fa-brands-github" />
  <div class="i-ph-lightning-duotone" />
</template>
```

You can use any icons from https://icones.js.org/

**Reply by ennioVisco** [reply]
> it's not work for me :( plz let me know how to add fontawesome to nuxt3
> 
> `/clients/plugins/fontawesome.ts`
> 
> ```
> iimport { defineNuxtPlugin } from 'nuxt3/app'
> 
> import { library, config } from '@fortawesome/fontawesome-svg-core'
> import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
> import { faArrowRotateRight } from '@fortawesome/free-solid-svg-icons'
> library.add(faArrowRotateRight)
> 
> config.autoAddCss = false
> 
> export default defineNuxtPlugin((nuxtApp) => {
>   nuxtApp.vueApp.component('FontAwesomeIcon', FontAwesomeIcon)
> })
> ```
> 
> `nuxt.config.ts`
> 
> ```
> import { defineNuxtConfig } from 'nuxt3'
> 
> export default defineNuxtConfig({
>   plugins: [
>     '~/clients/plugins/fontawesome.ts'
>   ],
>   build: {
>     transpile: [
>       '@fortawesome/vue-fontawesome',
>       '@fortawesome/fontawesome-svg-core',
>       '@fortawesome/free-regular-svg-icons',
>       '@fortawesome/free-solid-svg-icons',
>     ]
>   }
> })
> ```
> 
> `vue`
> 
> ```
> <template>
>   <div>
>     hello
>     <font-awesome-icon icon="fa-solid fa-arrow-rotate-right" />
>   </div>
> </template>
> ```
> 
> I have error message like this
> 
> ```
> [Vue warn]: Failed to resolve component: font-awesome-icon
> If this is a native custom element, make sure to exclude it from component resolution via compilerOptions.isCustomElement.
> [nitro] [dev] [unhandledRejection] TypeError: Invalid value used as weak map key
> [Vue warn]: Failed to resolve component: font-awesome-icon
> ```

It seems it cannot fetch the plugin. Why not making nuxt auto-import it and move it from `/clients/plugins` to `/plugins`. That way you could also remove the `plugins: [...]` field in the `nuxt.config.ts`.
Change the path and it should work.

**Reply by hsyoo1219** [reply]
I removed `plugins: [...]`, and it worked!
thanks everyone!

**Reply by wohui** [reply]
> You can use [unocss](https://github.com/unocss/unocss) more easily:
> 
> ```ts
> // nuxt.config.ts
> import presetIcons from '@unocss/preset-icons'
> import { defineNuxtConfig } from 'nuxt3'
> 
> export default defineNuxtConfig({
>   buildModules: ['@unocss/nuxt'],
> 
>   unocss: {
>     preflight: false,
>     presets: [
>       presetIcons(),
>     ],
>   },
> })
> ```
> 
> Then in any components:
> 
> ```
> <template>
>   <i class="i-fa-regular-heart" />
>   <div class="i-fa-brands-github" />
>   <div class="i-ph-lightning-duotone" />
> </template>
> ```
> 
> You can use any icons from https://icones.js.org/

why i got the icon height is 0 
nuxt3 like https://github.com/unocss/unocss

**Reply by mrc-bsllt** [reply]
> You can use [unocss](https://github.com/unocss/unocss) more easily:
> 
> ```ts
> // nuxt.config.ts
> import presetIcons from '@unocss/preset-icons'
> import { defineNuxtConfig } from 'nuxt3'
> 
> export default defineNuxtConfig({
>   buildModules: ['@unocss/nuxt'],
> 
>   unocss: {
>     preflight: false,
>     presets: [
>       presetIcons(),
>     ],
>   },
> })
> ```
> 
> Then in any components:
> 
> ```
> <template>
>   <i class="i-fa-regular-heart" />
>   <div class="i-fa-brands-github" />
>   <div class="i-ph-lightning-duotone" />
> </template>
> ```
> 
> You can use any icons from https://icones.js.org/

Hi stafyniaksacha, I copied your code, but I do not see the icons:
package.json
```
{
  "private": true,
  "scripts": {
    "build": "nuxt build",
    "dev": "nuxt dev",
    "generate": "nuxt generate",
    "preview": "nuxt preview"
  },
  "devDependencies": {
    "nuxt": "3.0.0-rc.4"
  },
  "dependencies": {
    "@nuxtjs/tailwindcss": "^5.1.3",
    "@pinia/nuxt": "^0.1.9",
    "@unocss/nuxt": "^0.43.2",
    "pinia": "^2.0.14",
    "sass": "^1.53.0",
    "sass-loader": "^13.0.2"
  }
}
```
nuxt.config.ts
```
import { defineNuxtConfig } from 'nuxt'
import presetIcons from '@unocss/preset-icons'

// https://v3.nuxtjs.org/api/configuration/nuxt.config
export default defineNuxtConfig({
  css: ["@/assets/scss/tailwind.css", "@/assets/scss/main.scss"],
  modules: ['@nuxtjs/tailwindcss', '@pinia/nuxt'],
  buildModules: ['@unocss/nuxt'],
  unocss: {
    preflight: false,
    presets: [
      presetIcons()
    ],
  }
})
```
index.vue
```
<template>
  <div>
    <p>{{ user?.username }}</p>
    <i class="i-fa-regular-heart" />
    <div class="i-fa-brands-github" />
    <div class="i-ph-lightning-duotone" />
  </div>
</template>

<script setup lang="ts">
import { API_HOST } from '@/utils/config'
import { logout } from '@/composables/logout'
import type { User } from '@/types/user'

const { data: user } = await useAsyncData<User | never>('user', () => {
  const token = useCookie('token').value
  const user_id = useCookie('user_id').value

  return $fetch(API_HOST + '/api/user/' + user_id, {
    headers: {
      Authorization: 'Bearer ' + token
    }
  })
})
if(!user.value) logout()
</script>
```
what am I missing? Thanks!

**Reply by stafyniaksacha** [reply]
> what am I missing? Thanks!

You have to install the icons sources package like `@iconify-json/[the-collection-you-want]` or `@iconify/json`
https://github.com/unocss/unocss/tree/main/packages/preset-icons#install

**Comment by QasimAliKhan027** [comment]
I am seeing this issue  "__vite_ssr_import_0__.default.component is not a function"

Here is code plugins/fontawesome.ts
import { defineNuxtPlugin } from 'nuxt3/app'

import { library, config } from '@fortawesome/fontawesome-svg-core'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import { faArrowRotateRight } from '@fortawesome/free-solid-svg-icons'
library.add(faArrowRotateRight)

config.autoAddCss = false

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.vueApp.component('FontAwesomeIcon', FontAwesomeIcon)
})

my nuxt.config.js look like this

import { defineNuxtConfig } from 'nuxt3'

// https://v3.nuxtjs.org/docs/directory-structure/nuxt.config
export default defineNuxtConfig({
  buildModules: [
    '@nuxtjs/tailwindcss',
    '@pinia/nuxt',
    '@nuxtjs/fontawesome'
  ],
  build: {
    transpile: [
      '@fortawesome/vue-fontawesome',
      '@fortawesome/fontawesome-svg-core',
      '@fortawesome/pro-solid-svg-icons',
      '@fortawesome/pro-regular-svg-icons',
      '@fortawesome/free-brands-svg-icons'
    ]
  }....
  .vue file
<template>
       <font-awesome-icon icon="fa-solid fa-arrow-rotate-right" />
</template>

**Comment by shtse8** [comment]
```
import { library } from '@fortawesome/fontawesome-svg-core'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import { faSms } from '@fortawesome/free-solid-svg-icons'
import { faFacebook, faTwitter } from '@fortawesome/free-brands-svg-icons'

export default defineNuxtPlugin((nuxtApp) => {
  console.log('[Plugin]', 'Font Awesome')
  library.add(faTwitter)
  library.add(faFacebook)
  library.add(faSms)
  nuxtApp.vueApp.component('font-awesome-icon', FontAwesomeIcon)
})
```

**Comment by cawa-93** [comment]
You don't need any special plugins. I just publish [iconify-prerendered](https://github.com/cawa-93/iconify-prerendered). You can casually import any icon as standalone vue component
```vue
<script setup>
import { IconSms } from '@iconify-prerendered/vue-fa-solid' 
</script>

<template>
  <IconSms /> <!-- will be rendered to raw <svg/> -->
</template>
```

**Comment by rodrigoodev** [comment]
I followed the suggestions and it worked for me

```js
// package.json
"dependencies": {
    "@fortawesome/fontawesome-svg-core": "^6.1.1",
    "@fortawesome/free-brands-svg-icons": "^6.1.1",
    "@fortawesome/free-regular-svg-icons": "^6.1.1",
    "@fortawesome/free-solid-svg-icons": "^6.1.1",
    "@fortawesome/vue-fontawesome": "^3.0.0-5"
  }
```

```js
//plugins/fontawesome.js
import { library, config } from "@fortawesome/fontawesome-svg-core";
import { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";
import { faUserSecret } from "@fortawesome/free-solid-svg-icons";

library.add(faUserSecret);

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.vueApp.component("font-awesome-icon", FontAwesomeIcon);
});
```

```js
//nuxt.config.ts
import { defineNuxtConfig } from "nuxt";

export default defineNuxtConfig({
  css: ["~/assets/css/main.css"],
  build: {
    transpile: [
      "@fortawesome/fontawesome-svg-core",
      "@fortawesome/free-brands-svg-icons",
      "@fortawesome/free-regular-svg-icons",
      "@fortawesome/free-solid-svg-icons",
      "@fortawesome/vue-fontawesome",
    ],
  },
});
```
```vue
<!-- app.vue -->
<template>
  <div>
    <font-awesome-icon :icon="['fa-solid', 'fa-user-secret']" />
  </div>
</template>
```

Attention to the imports version of fortawesome and vuefontawesome.

Unlike the people above, I didn't import the defineNuxtPlugin, and I wrote the export code without pasting, so it worked! There was some vscode import at this time of writing that made it work.

**Reply by cawa-93** [reply]
it may be so mutch more easier 
https://twitter.com/alex_kozack/status/1564289355161767936

**Comment by mzwoelfer** [comment]
Thanks. That worked for me.
My issue is, the next item after the `<font-awesome-icon icon="fas fa-trophy" />` is being displayed twice!

The span-tag containing the `navItem.name` is displayed twice, duplicating the text in my navigation bar
```
            <span class="p-1">
              <font-awesome-icon :icon="navItem.icon" />
              <span class="block">{{ navItem.name }}</span>
            </span>
```

**Comment by toniengelhardt** [comment]
Using FontAwesome icons (and other icons) becomes very easy with [nuxt-modules/icon](https://github.com/nuxt-modules/icon) package.
